Sara/sidebar v3.02/deployment (#1322)
semgrep/semgrep-docs: docs/deployment/deployment-checklist.md

---
slug: checklist
append_help_link: true
title: Pre-deployment checklist
description: tk
tags:
  - Deployment
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Before starting the deployment setup, use this checklist to ensure that:

- You and your organization agree on the **scope** of the deployment.
- You are aware of **permissions** that Semgrep needs to provide certain functions.
- You have **access** to the resources needed to carry out the deployment.

:::info
Ensure that your infrastructure meets all the [<i class="fa-regular fa-file-lines"></i> Prerequisites](/prerequisites) to run Semgrep.
:::

## Stakeholders and deployment team

> Semgrep integrates deeply and early in the development process. For medium-to-large teams, typically of more than 10 developers, coordinating with other departments before starting the deployment is crucial to an efficient roll-out.

Reviewer: ```suggestion
Semgrep integrates deeply and early in the development process. For medium-to-large teams, typically with more than 10 developers, coordinating with other departments before starting the deployment is crucial to an efficient roll-out.
```

> Semgrep integrates deeply and early in the development process. For medium-to-large teams, typically of more than 10 developers, coordinating with other departments before starting the deployment is crucial to an efficient roll-out.

Author: I reread this and realized that "early" is vague. Early as in before the developer commits the code? Or early as in "as soon as we buy the license? I rewrote the entire line.

Here are some teams or departments that may be responsible for parts of your Semgrep deployment:

| Department | Tasks related to deployment  |
| -------  | ------ |
| Infrastructure         | SSO, CI/CD, and source code manager (SCM) configuration.        |
| Engineering | Repository ownership, displaying findings to developers in PRs or MRs. |
| IT | Firewall or VPN configuration. |

## Scope

Scope refers to the breadth of deployment integration within your organization. The more users and repositories you onboard to Semgrep, the more crucial training becomes for **security champions** within your organization.

Ensure that all stakeholders agree on:

- Which users and departments will use Semgrep.
- Which repositories you will scan with Semgrep.
- How frequently you run Semgrep scans, such as daily or weekly, and at what time. This may affect other processes, such as PR approvals.
-  A timeframe for deployment. You may divide this into phases.

**Deployment times** vary greatly depending on your processes and size.

:::tip On scheduling scans

> Monorepos may take longer to finish scanning. Semgrep provides several options, including piecemeal scanning of the monorepo. See [<i class="fa-regular fa-file-lines"></i> Scan a large monorepo](/kb/semgrep-code/scanning-large-monorepo/) for more information.

Author: The file was renamed, I've updated. Thanks for the catch.

:::

## Roles

Semgrep provides two roles: `admin` and `member`.

For **single-user deployments**, you are the sole `admin` of your deployment.

For **multi-user deployments**, determine the following:

- The administrators (`admins`) that own the Semgrep deployment.
- For `members`, ensure that they have a sign-in method:
    - SSO
    - GitHub Cloud
    - GitLab Cloud

## Required permissions and access

The following checklist breaks down permissions required by Semgrep features.

<table>
<thead>
<tr>
<th>Feature</th>
<th>Permission required</th>
</tr>
</thead>
<tbody>
<tr>
<td rowspan="2">Run Semgrep continuously in your CI workflows.</td>
<td>Add or make changes to CI jobs. This includes committing configuration files for each</td>
</tr>
<tr>

> <td>Define environment variables and storing secrets.</td>

Author: For consistency, I changed all permissions to gerunds (-ing verbs) rather than simple English.

    values={[
    {label: 'Permissions for GitHub', value: 'permissions-github'},
    {label: 'Permissions for GitLab', value: 'permissions-gitlab'},
    ]}
>
<TabItem value='permissions-github'>

#### Permissions for GitHub

This section explains Semgrep Cloud Platform permissions that are requested in two different events:

* When you first sign in through GitHub.
* When you first add, integrate, or onboard your repositories to Semgrep Cloud Platform.

##### Permissions when signing in with GitHub

Semgrep Cloud Platform requests the following standard permissions set by GitHub when you first sign in. However, not all permissions are used by Semgrep Cloud Platform. Read the following list to see how Semgrep Cloud Platform uses permissions when signing in:

<dl>
    <dt>Verify your GitHub identity</dt>
    <dd>Enables Semgrep Cloud Platform to read your GitHub profile data, such as your username.</dd>
    <dt>Know which resources you can access</dt>
    <dd>Semgrep does not use or access any resources when first logging in. However, you can choose to share resources at a later point to add repositories into Semgrep Cloud Platform.</dd>
    <dt>Act on your behalf</dt>

>     <dd>Enables Semgrep Cloud Platform to perform certain tasks <strong>only on resources that you choose to share with Semgrep Cloud Platform</strong>. Semgrep Cloud Platform never uses this permission and never performs any actions on your behalf, even after you have installed <code>semgrep-app</code>. See <a href ="https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/authorizing-github-apps">When does a GitHub App act on your behalf?</a> in GitHub documentation.</dd>

Reviewer 2: I'm confused by the "Semgrep Cloud Platform never uses this permission" statement. We must be using it no?

Author: Hahaha I had the exact same question when I wrote this originally. When a user "authorizes" a github app, there are 3 permissions at the minimum that GitHub requires the user accept for ANY GitHub app. This is regardless of any actual need for those permissions. The link provided to the github docs states:

> Act on your behalf: When authorized, the application may perform tasks on GitHub on your behalf. This might include creating an issue or commenting on a pull request. For more information, see "[About GitHub Apps acting on your behalf](https://docs.github.com/en/apps/using-github-apps/authorizing-github-apps#about-github-apps-acting-on-your-behalf)."

You will notice that when a comment is posted, it is `semgrep-app` that performs the action, not the user. So `semgrep-app` is not acting on a user's behalf (or it would have used the user's account).

    <dt>Reading and writing 2 files</dt>
    <dd>Enables Semgrep Cloud Platform to configure itself to run in CI by writing to <code>.github/workflows/semgrep.yml</code> and <code>.semgrepignore</code> files.</dd>
    <dt>Reading and writing workflows</dt>
    <dd>Enables Semgrep Cloud Platform to configure itself to run in CI by writing to <code>.github/workflows/semgrep.yml</code>. GitHub allows writing to files within <code>.github/workflows/</code> directory only if this permission is granted along with "Writing a single file."</dd>
    <dt>Reading and writing pull requests</dt>
    <dd>Write permissions allow Semgrep Cloud Platform to leave pull request comments about findings. Read permissions allow Semgrep Cloud Platform to automatically remove findings when the pull request that introduced them is closed without merging.</dd>
</dl>

</TabItem>

<TabItem value='permissions-gitlab'>

#### Permissions for GitLab

Semgrep requires the following permissions (scopes) to enable the authentication of a session:

* `openid`
* `email`
* `profile`
* `API`

</TabItem>
</Tabs>

### Semgrep versions

> Many improvements to the Semgrep Cloud Platform experience only work with up-to-date Semgrep CLI versions. As such, Semgrep Cloud Platform only supports the 10 most recent minor versions of Semgrep CLI. For example, if the latest release was 0.114.0, all versions greater than 0.104.0 are supported, while earlier versions, such as 0.103.0, can be deprecated or can result in failures.

Reviewer 2: small nitpick, but can we consider using versions that are a bit less awkward? maybe 1.52.0 - 1.62.0? maybe this doesn't matter, not sure.

---
slug: checklist
append_help_link: true
title: Pre-deployment checklist
description: tk
tags:
  - Deployment
---

> Before starting the deployment set-up, use this checklist to ensure that:

Reviewer: ```suggestion
Before starting the deployment setup, use this checklist to ensure that:
```

- You have **access** to the resources needed to carry out the deployment.
- You and your organization agree on the **scope** of the deployment.

## Permissions and  access 

Ensure that you and your deployment team have sufficient permissions to:

- Add or make changes to CI jobs.
- Create access tokens in your source code manager.
- Commit files to all repositories you want Semgrep to scan.
- (Optional) View SSO configurations.
- (Optional) Set up channels in your chosen notification method (Slack, email, or webhooks).
- (For self-hosted repositories) Edit your firewall or VPN configuration's allowlist.

## Resources

> Determine the following:

Reviewer: ```suggestion
Ensure that you and your deployment team agree on:
```
This is a minor thing, but I thought a bit more clarity here would be helpful

- What roles or departments will use Semgrep.
- The number of repositories you will scan with Semgrep.

## Processes

> Determine the following:

Reviewer: Similar to my previous comment -- it might be helpful to have a bit more context here?

Author: I've redone the deployment checklist to make it an actual table checklist with better information, i hope it resolves this.

> - When you want to run a scan:

Reviewer: ```suggestion
- When you want to run a scan; common options include:
```
